bitcoin/examples/sighash.rs

```diff
@@ -0,0 +1,234 @@
+        script_pubkey_bytes = bytes;
+    } else {
+        //In the P2SH case we get scriptPubKey from scriptSig of the spending input.
+        //The scriptSig that corresponds to an M of N multisig should be: PUSHBYTES_0 PUSHBYTES_K0 <sig0><sighashflag0> ... PUSHBYTES_Km <sigM><sighashflagM> PUSHBYTES_X <scriptPubKey>
+        //Here we assume that we have an M of N multisig scriptPubKey.
+        for (k, instr) in script_sig.instructions().enumerate() {
+            match instr.unwrap() {
+                Instruction::PushBytes(pb) => {
+                    if k == 0 {
+                        assert!(
+                            pb.is_empty(),
+                            "first in ScriptSig must be PUSHBYTES_0 got {:?}",
+                            pb
+                        )
+                    } else if script_pubkey_bytes_opt.is_none()
+                        && k == script_sig.instructions().count() - 1
+                    {
+                        //In P2SH the last stack element from scriptSig is ScriptPubkey
+                        script_pubkey_bytes = pb.as_bytes();
+                    } else {
+                        //All others from 1 (0-indexed) to M-2 must be signatures
+                        let sig_bytes = pb.as_bytes();
+                        let sig = bitcoin::ecdsa::Signature::from_slice(sig_bytes)
+                            .expect("failed to parse sig");
+                        let sig_len = sig_bytes.len() - 1; //last byte is EcdsaSighashType sighash flag
+                                                           //ECDSA signature in DER format lengths are between 70 and 72 bytes
+                        assert!(
+                            (70..=72).contains(&sig_len),
+                            "signature length {} out of bounds",
+                            sig_len
+                        );
+                        //take sighash_flag into account - can they be different?
+                        assert!(
+                            last_sighash_flag == bitcoin::sighash::EcdsaSighashType::None
+                                || last_sighash_flag == sig.hash_ty,
+                            "different sighash flags"
+                        );
+                        last_sighash_flag = sig.hash_ty;
+                    }
+                    println!("ScriptSig PushBytes len {}: {:?}", pb.as_bytes().to_vec().len(), pb)
+                }
+                Instruction::Op(op) => {
+                    unreachable!("we only expect PushBytes here, got Op({})", op)
+                }
+            }
+        }
+    }
+
+    let script_code = bitcoin::Script::from_bytes(script_pubkey_bytes);
+    let cache = sighash::SighashCache::new(&tx);
+    let sighash = cache
+        .legacy_signature_hash(inp_idx, script_code, last_sighash_flag.to_u32())
+        .expect("failed to compute sighash");
+    println!("Legacy sighash: {:x}\n", sighash);
```

Comment:
if `script_pubkey_bytes_opt` was `None` this will use an empty slice for `script_code` and the sighash flag from the last signature even though the input index is taken as a parameter. I don't think this is correct but I could be confused?

Perhaps it would be better to split this function into two so things like `script_code` aren't overloaded by different use cases - these variables (and concepts) are super hard to keep clear at the best of times (for me at least).

Please take everything I said with a pinch of salt. In case its useful here is a blogpost I've read a 100 times to try to understand all the sighash types: https://raghavsood.com/blog/2018/06/10/bitcoin-signature-types-sighash

Reply:
> if `script_pubkey_bytes_opt` was `None` this will use an empty slice for `script_code` and the sighash flag from the last signature even though the input index is taken as a parameter. I don't think this is correct but I could be confused?
> 
> Perhaps it would be better to split this function into two so things like `script_code` aren't overloaded by different use cases - these variables (and concepts) are super hard to keep clear at the best of times (for me at least).
> 
> Please take everything I said with a pinch of salt. In case its useful here is a blogpost I've read a 100 times to try to understand all the sighash types: https://raghavsood.com/blog/2018/06/10/bitcoin-signature-types-sighash

> Either I'm not understanding things fully or there is too much going on in this function (see comment below).

`compute_sighash_legacy()` is exactly the same function for  both P2MS and P2SH multisig, except that for P2MS, the last parameter, `scriptPubKey` is in the referenced output, therefore it has to be supplied externally from the original transaction, while in P2SH it's already in the input's scriptSig of the spending transaction that is the first parameter to `compute_sighash_legacy()`, so it can be extracted from the spending transaction, so we pass None for `scriptPubKey`.  
Perhaps I can clarify that with a comment?

PS. I could reproduce and fixed the 1.48 toolchain problem as you suggested, thanks, sry i messed up my playground while switching toolchains.

Reply:
Yep I got all that :) If the others rekon its good then its good for me. Thanks